tempesta_fw/str_avx2.S

```diff
@@ -0,0 +1,2296 @@
+	 * If the tail is shorter than 8 bytes, revert the string pointer back
+	 * to get 8 bytes for 1 vector iteration processing.
+	 */
+.stricmp_tail:
+	movslq	%eax, %r8
+	cmpq	%r8, %rdx
+	je	.stricmp_match
+	subq	%r8, %rdx
+	vmovdqa	__D, %xmm3
+	leal	-8(%rax,%rdx), %ecx
+	cmpq	$7, %rdx
+	movslq	%ecx, %rcx
+	cmova	%r8, %rcx
+	vmovsd	-8(%rdi,%rcx), %xmm1
+	vmovsd	-8(%rsi,%rcx), %xmm0
+	vmovhpd	(%rdi,%rcx), %xmm1, %xmm1
+	vmovhpd	(%rsi,%rcx), %xmm0, %xmm0
+	vpxor	%xmm0, %xmm1, %xmm2
+	vmovdqa	__CASE, %xmm0
+	vpor	%xmm1, %xmm0, %xmm1
+	vpsubb	__a, %xmm1, %xmm1
+	vpcmpgtb %xmm1, %xmm3, %xmm1
+	vpcmpeqb %xmm0, %xmm2, %xmm3
+	vpand	%xmm3, %xmm0, %xmm0
+	vpand	%xmm1, %xmm0, %xmm0
+	vpxor	%xmm1, %xmm1, %xmm1
+	vpxor	%xmm2, %xmm0, %xmm0
+	vpcmpeqb %xmm1, %xmm0, %xmm0
+	vpmovmskb %xmm0, %eax
+	xorl	$0xffff, %eax
+	ret
+
+.stricmp_match:
+	xorl	%eax, %eax
+	ret
+
+.stricmp_128:
+	leaq	(%rsi,%rax), %r8
+	addq	%rdi, %rax
+	vmovdqa	__a, %ymm6
+	vlddqu	96(%rax), %ymm5
+	vlddqu	96(%r8), %ymm0
+	vlddqu	(%rax), %ymm11
+	vpxor	%ymm0, %ymm5, %ymm4
+	vlddqu	(%r8), %ymm1
+	vmovdqa	__CASE, %ymm0
+	vpxor	%ymm1, %ymm11, %ymm12
+	vmovdqa	__D, %ymm3
+	vpor	%ymm11, %ymm0, %ymm11
+	vlddqu	32(%rax), %ymm10
+	vlddqu	32(%r8), %ymm9
+	vpor	%ymm5, %ymm0, %ymm5
+	vlddqu	64(%rax), %ymm8
+	movslq	%ecx, %rax
+	vpsubb	%ymm6, %ymm11, %ymm1
+	vpcmpgtb %ymm1, %ymm3, %ymm11
+	vpxor	%ymm9, %ymm10, %ymm9
+	vpor	%ymm10, %ymm0, %ymm10
+	vlddqu	64(%r8), %ymm7
+	vpcmpeqb %ymm0, %ymm12, %ymm1
+	vpxor	%ymm7, %ymm8, %ymm7
+	vpor	%ymm8, %ymm0, %ymm8
+	vpand	%ymm1, %ymm0, %ymm1
+	vpand	%ymm11, %ymm1, %ymm1
+	vpxor	%ymm12, %ymm1, %ymm1
+	vpcmpeqb %ymm2, %ymm1, %ymm1
+	vpmovmskb %ymm1, %r9d
+	vpsubb	%ymm6, %ymm10, %ymm1
+	vpcmpgtb %ymm1, %ymm3, %ymm10
+	movl	%r9d, %ecx
+	vpcmpeqb %ymm9, %ymm0, %ymm1
+	vpand	%ymm1, %ymm0, %ymm1
+	vpand	%ymm10, %ymm1, %ymm1
+	vpxor	%ymm9, %ymm1, %ymm1
+	vpcmpeqb %ymm2, %ymm1, %ymm1
+	vpmovmskb %ymm1, %r11d
+	vpsubb	%ymm6, %ymm8, %ymm1
+	vpsubb	%ymm6, %ymm5, %ymm6
+	vpcmpgtb %ymm1, %ymm3, %ymm8
+	andl	%r11d, %ecx
+	vpcmpeqb %ymm7, %ymm0, %ymm1
+	vpcmpgtb %ymm6, %ymm3, %ymm3
+	vpand	%ymm1, %ymm0, %ymm1
+	vpand	%ymm8, %ymm1, %ymm1
+	vpxor	%ymm7, %ymm1, %ymm1
+	vpcmpeqb %ymm2, %ymm1, %ymm1
+	vpmovmskb %ymm1, %r10d
+	vpcmpeqb %ymm4, %ymm0, %ymm1
+	andl	%r10d, %ecx
+	vpand	%ymm1, %ymm0, %ymm0
+	vpand	%ymm3, %ymm0, %ymm0
+	vpxor	%ymm4, %ymm0, %ymm0
+	vpcmpeqb %ymm2, %ymm0, %ymm0
+	vpmovmskb %ymm0, %r8d
+	andl	%ecx, %r8d
+	cmpl	$~0, %r8d
+	je	.stricmp_128_loop
+
+.stricmp_nomatch:
+	movl	$1, %eax
+	ret
+
+.stricmp_64:
+	leaq	(%rsi,%rax), %r8
+	addq	%rdi, %rax
+	vmovdqa	__a, %ymm7
+	vlddqu	32(%rax), %ymm3
+	vlddqu	32(%r8), %ymm0
+	vlddqu	(%rax), %ymm1
+	vpxor	%ymm0, %ymm3, %ymm4
+	vlddqu	(%r8), %ymm6
+	vmovdqa	__CASE, %ymm0
+	vpxor	%ymm6, %ymm1, %ymm6
+	vmovdqa	__D, %ymm5
+	vpor	%ymm1, %ymm0, %ymm1
+	vpsubb	%ymm7, %ymm1, %ymm1
+	vpcmpgtb %ymm1, %ymm5, %ymm2
+	vpcmpeqb %ymm0, %ymm6, %ymm1
+	vpand	%ymm1, %ymm0, %ymm1
+	vpand	%ymm2, %ymm1, %ymm1
+	vpxor	%xmm2, %xmm2, %xmm2
```

Comment:
This looks out of place. Other instructions are working with `ymm` registers, but this one only clears `xmm2`, lower part of `ymm2`. Why?

Reply:
Oh, I forgot that VEX-encoded instructions operating on `xmm` reset upper half of `ymm` register, so this effectively zeroes the whole `ymm2`.